File: HouseRobber2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Nov 27 08:45:29 2019

@author: tanvirkaur
"""

#time complexity = O(n)
#Space complexity = O(n*2)
#LeetCode acceptance = Yes
# using 2D matrix

class Solution:
    def rob(self, nums: List[int]) -> int:
        # base case
        if not nums:
            return 0
        dp = [[0 for _ in range(2)] for _ in range(len(nums))]
        dp[0][1] = nums[0]
        for i in range(1,len(nums)):
            # if we not choose
            dp[i][0] = max(dp[i-1][0], dp[i-1][1])
            # if we choose we have to take the amount of the previous not choosen
            dp[i][1] = nums[i] + dp[i-1][0]
            # in the end we return the maximum of the last row
        return max(dp[len(nums)-1][0],dp[len(nums)-1][1])

File: HouseRobber2.java
// Time Complexity : O(n) , n = length of nums array
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
class Solution {
    public int rob(int[] nums) {
        if(nums == null || nums.length == 0)
            return 0;
        
        if(nums.length == 1)
            return nums[0];
        
        int choose = nums[0], skip = 0;
        for(int i=1; i<nums.length; i++) {
            int prevSkip = skip;
            skip = Math.max(choose, skip);
            choose = nums[i] + prevSkip;
        }
        return Math.max(choose,skip);
    }
}